erdgo/wallet.go

```diff
@@ -0,0 +1,136 @@
+package erdgo
+
+import (
+	"crypto/hmac"
+	"crypto/sha512"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/pem"
+	"io/ioutil"
+	"os"
+
+	"github.com/ElrondNetwork/elrond-go/crypto/signing"
+	"github.com/ElrondNetwork/elrond-go/crypto/signing/ed25519"
+	"github.com/tyler-smith/go-bip39"
+)
+
+const (
+	egldCoinType = uint32(508)
+	hardened     = uint32(0x80000000)
+)
+
+type bip32Path []uint32
+
+type bip32 struct {
+	Key       []byte
+	ChainCode []byte
+}
+
+var egldPath = bip32Path{
+	44 | hardened,
+	egldCoinType | hardened,
+	hardened, // account
+	hardened,
+	hardened, // addressIndex
+}
+
+// GenerateNewMnemonic generates a new set of 24 words to be used as seed phrase
+func GenerateNewMnemonic() (string, error) {
+	entropy, err := bip39.NewEntropy(256)
+	if err != nil {
+		return "", err
+	}
+
+	return bip39.NewMnemonic(entropy)
+}
+
+// GetPrivateKeyFromMnemonic generates a private key based on mnemonic, account and address index
+func GetPrivateKeyFromMnemonic(mnemonic string, account, addressIndex uint8) []byte {
+	seed := bip39.NewSeed(mnemonic, "")
+	egldPath[2] = uint32(account) | hardened
+	egldPath[4] = uint32(addressIndex) | hardened
+	keyData := derivePrivateKey(seed, egldPath)
+
+	return keyData.Key
+}
+
+// GetAddressFromPrivateKey generates the bech32 address from a private key
+func GetAddressFromPrivateKey(privateKeyBytes []byte) (string, error) {
+	suite := ed25519.NewEd25519()
+	keyGen := signing.NewKeyGenerator(suite)
+	privateKey, err := keyGen.PrivateKeyFromByteArray(privateKeyBytes)
+	if err != nil {
+		return "", err
+	}
+	publicKey := privateKey.GeneratePublic()
+	publicKeyBytes, err := publicKey.ToByteArray()
+	if err != nil {
+		return "", err
+	}
+
+	return PubkeyToBech32(publicKeyBytes)
+}
+
+// LoadPrivateKeyFromPemFile loads a private key from a .pem file
+func LoadPrivateKeyFromPemFile(filename string) ([]byte, error) {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	blk, _ := pem.Decode(data)
+	if blk == nil {
+		return nil, errInvalidPemFile
+	}
+
+	return hex.DecodeString(string(blk.Bytes))
+}
+
+// SavePrivateKeyToPemFile saves the private key in a .pem file
+func SavePrivateKeyToPemFile(privateKey []byte, filename string) error {
+	address, err := GetAddressFromPrivateKey(privateKey)
+	if err != nil {
+		return err
+	}
+	if len(privateKey) <= pubkeyLen {
+		pubkey, err := Bech32ToPubkey(address)
+		if err != nil {
+			return err
+		}
+		privateKey = append(privateKey, pubkey...)
+	}
+	blk := pem.Block{
+		Type:  "PRIVATE KEY for " + address,
+		Bytes: []byte(hex.EncodeToString(privateKey)),
+	}
+	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = file.Close()
+	}()
+
+	return pem.Encode(file, &blk)
+}
+
+func derivePrivateKey(seed []byte, path bip32Path) *bip32 {
+	b := &bip32{}
+	digest := hmac.New(sha512.New, []byte("ed25519 seed"))
```

Comment:
can you have the string for the parent chain as a parameter? ("ed25519 seed") should come as parameter

this is because for mainnet for example we have empty string for the parent chain so ""
and if you want to create the same (private key, pub key pair) from erdgo compatible with mainnet then you would give empty string. 
 
you can also have it default to ""

Reply:
I would rather leave "ed25519 seed" for now to maintain compatibility with our Ledger app 🙏

Reply:
@AdoAdoAdo, I can confirm, this is somehow part of `bip32`, and we use it in `core-js`, `erdpy` and `erdjava` as well:

```
elrond-core-js/node_modules/ed25519-hd-key/dist/index.js
elrond-sdk/erdjava/src/main/java/elrond/Wallet.java
elrond-sdk/erdpy/wallet/core.py
```